### DISCUSSION TITLE
Setting up the i2c_bitbang functions in zephyr/drivers/i2c/i2c_bitbang.cpp

### DISCUSSION THREAD

**Discussion by michel-ann** [zephyrproject-rtos/zephyr#39049] [Q&A]
I am trying to use the functions outlined by this file, but I am a little confused about how to initialize everything. All of the functions use a **struct i2c_bitbang *context**, which is the struct given in the header file. My question is, how should I set up the struct? I am just confused on what it wants for the **void *io_context** field and how to declare the second **struct i2c_bitbang_io *io** field.
Any help / clarification would be much appreciated. Thank you!

**Comment by yashi** [comment]
The file `drivers/i2c/i2c_bitbang.c` is a low level driver.  You don't directly use it.  Rather, you should use API provided by `include/drivers/i2c.h`.

So to answer your question:
>  how should I set up the [struct i2c_bitbang *context] struct?

You don't.   High level drivers, such as `drivers/i2c/i2c_gpio.c`, does.

```
$ git grep i2c_bitbang_init
drivers/i2c/i2c_bitbang.c:void i2c_bitbang_init(struct i2c_bitbang *context,
drivers/i2c/i2c_bitbang.h:void i2c_bitbang_init(struct i2c_bitbang *bitbang,
drivers/i2c/i2c_gpio.c: i2c_bitbang_init(&context->bitbang, &io_fns, context);
drivers/i2c/i2c_litex.c:        i2c_bitbang_init(bitbang, &i2c_litex_bitbang_io, (void *)config);
drivers/i2c/i2c_sbcon.c:        i2c_bitbang_init(&context->bitbang, &io_fns, config->sbcon);
```
